mempool/src/mempool.rs

```diff
@@ -754,39 +1053,52 @@ impl MempoolState {
         tx_hash: &Blake2bHash,
         reason: EvictionReason,
     ) -> Option<Transaction> {
-        let tx = self.transactions.remove(tx_hash)?;
+        if let Some(tx) = self.regular_transactions.delete(tx_hash) {
+            let sender_state = self.state_by_sender.get_mut(&tx.sender).unwrap();
 
-        self.best_transactions.remove(tx_hash);
-        self.worst_transactions.remove(tx_hash);
-        self.oldest_transactions.remove(tx_hash);
+            sender_state.total -= tx.total_value();
+            sender_state.txns.remove(tx_hash);
 
-        let sender_state = self.state_by_sender.get_mut(&tx.sender).unwrap();
+            if sender_state.txns.is_empty() {
+                self.state_by_sender.remove(&tx.sender);
+            }
 
-        sender_state.total -= tx.total_value();
-        sender_state.txns.remove(tx_hash);
+            self.remove_from_staking_state(&tx);
 
-        if sender_state.txns.is_empty() {
-            self.state_by_sender.remove(&tx.sender);
-        }
+            #[cfg(feature = "metrics")]
+            self.metrics.note_evicted(reason);
 
-        self.remove_from_staking_state(&tx);
+            return Some(tx);
+        } else if let Some(tx) = self.control_transactions.delete(tx_hash) {
+            let sender_state = self.state_by_sender.get_mut(&tx.sender).unwrap();
 
-        self.total_size -= tx.serialized_size();
+            sender_state.total -= tx.total_value();
+            sender_state.txns.remove(tx_hash);
 
-        #[cfg(feature = "metrics")]
-        self.metrics.note_evicted(reason);
+            if sender_state.txns.is_empty() {
+                self.state_by_sender.remove(&tx.sender);
+            }
 
-        Some(tx)
+            self.remove_from_staking_state(&tx);
+
+            #[cfg(feature = "metrics")]
+            self.metrics.note_evicted(reason);
+
+            return Some(tx);
+        } else {
+            return None;
+        }
     }
 
     // Removes all the transactions sent by some specific address
     pub(crate) fn remove_sender_txns(&mut self, sender_address: &Address) {
         if let Some(sender_state) = &self.state_by_sender.remove(sender_address) {
             for tx_hash in &sender_state.txns {
-                self.best_transactions.remove(tx_hash);
-                self.worst_transactions.remove(tx_hash);
-                self.oldest_transactions.remove(tx_hash);
-                if let Some(tx) = self.transactions.remove(tx_hash) {
+                if let Some(tx) = self.regular_transactions.delete(tx_hash) {
+                    self.remove_from_staking_state(&tx);
+                }
+
+                if let Some(tx) = self.control_transactions.delete(tx_hash) {
                     self.remove_from_staking_state(&tx);
                 }
```

Comment:
Similar to above, can this be combined?
```rust
if let Some(tx) = self.regular_transactions.delete(tx_hash) || self.control_transactions.delete(tx_hash) {
    self.remove_from_staking_state(&tx);
}
```

Reply:
Same as above, forbidden. 
Probably the reason is: what happens if both regular_transactions and control_transactions return an object?
if they both return an object, tx would be assigned to what? to the object returned by the regular_transactions or the one returned by control_transactions?

Reply:
In Javascript, the functions are run after each other. If the first call returns a truthy value, the second function is not called anymore. That's what I would expect. And in this case it makes sense, because only one tx bucket can contain the tx_hash in this implementation.

Reply:
Just saw this.

`||` does not work, but `or_else(...)` should.

```suggestion
                if let Some(tx) = self.regular_transactions.delete(tx_hash)
                    .or_else(|| self.control_transactions.delete(tx_hash)){ 
                    self.remove_from_staking_state(&tx);
                }
```